src/igraph_stubs.jl

```diff
@@ -0,0 +1,332 @@
+#### Optional APIs backed by external packages (e.g., IGraphs.jl)
+#### These functions are PUBLIC names that Graphs.jl exposes, but by default
+#### they throw a helpful error until a backend (IGraphs.jl) provides methods.
+
+# Friendly message builder
+_igraph_backend_msg(fname) = """
+`$(fname)` is not implemented in base Graphs.jl.
+It is available via the **IGraphs.jl** backend.
+
+Enable it and re-run:
+
+    using IGraphs, IGraphs.GraphsCompat
+    $(fname)(g, IGraphAlgorithm())
+
+"""
+
+# -----------------------
+# Assortativity
+# -----------------------
+@static if !isdefined(@__MODULE__, :assortativity)
+"""
+    assortativity(g::AbstractGraph, attr; normalized::Bool=true)
+
+Degree- or attribute-based assortativity coefficient.
+
+!!! note "Availability"
+    Backend-only: use **IGraphs.jl** (see message if not loaded).
+"""
+function assortativity end
+assortativity(g::AbstractGraph, attr; kwargs...) =
+    throw(ArgumentError(_igraph_backend_msg(:assortativity)))
+end
+
+@static if !isdefined(@__MODULE__, :assortativity_degree)
+"""
+    assortativity_degree(g::AbstractGraph; normalized::Bool=true)
+
+Degree assortativity (scalar coefficient).
+"""
+function assortativity_degree end
+assortativity_degree(g::AbstractGraph; kwargs...) =
+    throw(ArgumentError(_igraph_backend_msg(:assortativity_degree)))
+end
+
+@static if !isdefined(@__MODULE__, :assortativity_nominal)
+"""
+    assortativity_nominal(g::AbstractGraph, attr)
+
+Nominal/categorical assortativity based on a vertex attribute.
+"""
+function assortativity_nominal end
+assortativity_nominal(g::AbstractGraph, attr; kwargs...) =
+    throw(ArgumentError(_igraph_backend_msg(:assortativity_nominal)))
```

Comment:
This probably should just throw the default MethodError, not have a custom error defined for each method. A big drawback of the current approach you have here is the significant extra compilation and then invalidation on importing other packages.

In julia it is idiomatic to rely on MethodErrors to mark unimplemented APIs. Then you can add error hints to be more helpful to the user. All of these should probably be converted to error hints, but let's leave that to another PR -- for the moment we can just focus on removing all these default methods.